service/src/main/java/bio/terra/tanagra/app/controller/CohortsV2ApiController.java

```diff
@@ -0,0 +1,204 @@
+    accessControlService.throwIfUnauthorized(null, DELETE, COHORT, new ResourceId(cohortId));
+    cohortService.deleteCohort(studyId, cohortId);
+    return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+  }
+
+  @Override
+  public ResponseEntity<ApiCohortV2> getCohort(String studyId, String cohortId) {
+    accessControlService.throwIfUnauthorized(null, READ, COHORT, new ResourceId(cohortId));
+    return ResponseEntity.ok(toApiObject(cohortService.getCohort(studyId, cohortId)));
+  }
+
+  @Override
+  public ResponseEntity<ApiCohortListV2> listCohorts(
+      String studyId, Integer offset, Integer limit) {
+    ResourceIdCollection authorizedCohortIds =
+        accessControlService.listResourceIds(COHORT, offset, limit);
+    List<Cohort> authorizedCohorts;
+    if (authorizedCohortIds.isAllResourceIds()) {
+      authorizedCohorts = cohortService.getAllCohorts(studyId, offset, limit);
+    } else {
+      authorizedCohorts =
+          cohortService.getCohorts(
+              studyId,
+              authorizedCohortIds.getResourceIds().stream()
+                  .map(ResourceId::getId)
+                  .collect(Collectors.toList()),
+              offset,
+              limit);
+    }
+
+    ApiCohortListV2 apiCohorts = new ApiCohortListV2();
+    authorizedCohorts.stream().forEach(cohort -> apiCohorts.add(toApiObject(cohort)));
+    return ResponseEntity.ok(apiCohorts);
+  }
+
+  @Override
+  public ResponseEntity<ApiCohortV2> updateCohort(
+      String studyId, String cohortId, ApiCohortUpdateInfoV2 body) {
+    accessControlService.throwIfUnauthorized(null, UPDATE, COHORT, new ResourceId(cohortId));
+
+    List<CriteriaGroup> criteriaGroups =
+        body.getCriteriaGroups() == null
+            ? null
+            : body.getCriteriaGroups().stream()
+                .map(
+                    apiCriteriaGroup -> {
+                      // Generate random 10-character alphanumeric strings for each criteria group
+                      // and criteria (below). We maintain 2 separate IDs (id and user_facing_id) so
+                      // that the UI-specified id doesn't need to match the DB schema on the backend
+                      // (e.g. for uniqueness). Right now, the API doesn't allow callers to update
+                      // individual criteria groups or criteria, just the whole cohort. This means
+                      // we can simplify the current implementation and just delete/re-insert all
+                      // criteria groups and criteria every time the cohort gets updated, rather
+                      // than diffing the criteria groups and updating/inserting/deleting. This is
+                      // why the IDs are randomly generated on each update. If we move away from the
+                      // simple implementation, then we'll have to expose the internal ids through
+                      // the service API.
+                      String newCriteriaGroupId = RandomStringUtils.randomAlphanumeric(15);
```

Comment:
Is there a reason this is 15 instead of 10 like above? This is also excessive. It's a larger key space than a 64 bit number when we could easily be using 32 bit keys (~6 alphanumeric characters).

Reply:
Ah good catch. No, I meant to make them the same.